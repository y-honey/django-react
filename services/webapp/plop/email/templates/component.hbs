import { FormattedMessage } from 'react-intl';
import { EmailComponentProps } from '../../types';
import { Button, Layout } from '../../base';
import { useGenerateAbsoluteLocalePath } from '../../../shared/hooks/localePaths';

export type {{ pascalCase name }}Props = EmailComponentProps & {};

export const Template = ({}: {{ pascalCase name }}Props) => {
  const generateLocalePath = useGenerateAbsoluteLocalePath();
  const url = generateLocalePath(['home']);

  return (
    <Layout
      title={<FormattedMessage defaultMessage="Title" id="Email / {{ titleCase name }} / Title" />}
      text={<FormattedMessage defaultMessage="Text" id="Email / {{ titleCase name }} / Text" />}
    >
      <Button linkTo={url}>
        <FormattedMessage defaultMessage="Button" id="Email / {{ titleCase name }} / Link label" />
      </Button>
    </Layout>
  );
};

export const Subject = () => <FormattedMessage defaultMessage="Subject" id="Email / {{ titleCase name }} / Subject" />
