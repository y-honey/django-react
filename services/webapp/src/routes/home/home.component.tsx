import React from 'react';
import { Helmet } from 'react-helmet-async';
import { FormattedMessage, useIntl } from 'react-intl';

import { LanguageSwitcher } from '../../shared/components/languageSwitcher';
import { H1, H2 } from '../../theme/typography';
import { LoginForm } from '../../shared/components/auth/loginForm';
import { SignupForm } from '../../shared/components/auth/signupForm';
import { Container, Logo } from './home.styles';

export const Home = () => {
  const intl = useIntl();

  return (
    <Container>
      <Helmet
        title={intl.formatMessage({
          defaultMessage: 'Homepage',
          description: 'Home / page title',
        })}
      />

      <H1>
        <FormattedMessage defaultMessage="Hello world!" description="Home / title" />
      </H1>

      <Logo />

      <LanguageSwitcher />

      <H2>
        <FormattedMessage defaultMessage="Register" description="Home / register" />
      </H2>
      <SignupForm />

      <H2>
        <FormattedMessage defaultMessage="Login" description="Home / login" />
      </H2>
      <LoginForm />
    </Container>
  );
};
