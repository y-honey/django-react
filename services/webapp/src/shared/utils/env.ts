export const getEnvNumber = (key: keyof NodeJS.ProcessEnv) => {
  const value = process.env[key];
  if (value === undefined) return;
  const number = Number(value);
  if (isNaN(number)) return;
  return number;
};
