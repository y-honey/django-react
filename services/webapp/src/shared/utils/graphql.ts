import { UnknownObject } from './types';

export type ConnectionType<ITEM> = {
  readonly edges: ReadonlyArray<{
    readonly node: ITEM | null;
  } | null>;
} | null;

export const mapConnection = <ITEM, RETURN>(
  callback: (item: ITEM) => RETURN,
  query: ConnectionType<ITEM>
): RETURN[] => {
  const existingNodes: RETURN[] = [];
  const edges = query?.edges ?? [];

  for (const edge of edges) {
    if (edge?.node) {
      existingNodes.push(callback(edge.node));
    }
  }

  return existingNodes;
};

export type ExtractNodeType<
  T extends {
    readonly edges: ReadonlyArray<{
      readonly node: ITEM | null;
    } | null>;
  } | null,
  ITEM extends UnknownObject = UnknownObject
> = NonNullable<NonNullable<NonNullable<NonNullable<T>['edges']>[number]>['node']>;
