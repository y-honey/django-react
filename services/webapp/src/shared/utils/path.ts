import { map } from 'ramda';

export const getLocalePath = (p: string) => `/:lang/${p}`;

export const nestedPath = <T extends string>(root: string, nestedRoutes: Record<T, string>) => {
  const absoluteNestedUrls = map<Record<T, string>, Record<T, string>>((value) => root + '/' + value, nestedRoutes);
  const paths = {
    index: `${root}/*`,
    ...absoluteNestedUrls
  };
  return {
    ...paths,
    getRelativeUrl: (route: T) => nestedRoutes[route],
    getLocalePath: (route: T) => getLocalePath(paths[route]),
  };
};
