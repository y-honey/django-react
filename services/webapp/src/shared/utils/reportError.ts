import * as Sentry from '@sentry/react';

/**
 * Generic function for error reporting.
 */
export const reportError = (error: any): void => {
  if (process.env.NODE_ENV !== 'test') {
    console.error(error); //eslint-disable-line
  }
  if (!error?.response) {
    Sentry.captureException(error);
  }
};
