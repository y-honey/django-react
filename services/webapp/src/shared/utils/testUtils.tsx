import { ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Route, Routes } from 'react-router';
import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { produce } from 'immer';
import { RelayEnvironmentProvider } from 'react-relay';
import { createMockEnvironment, RelayMockEnvironment } from 'relay-test-utils';

import { DEFAULT_LOCALE, translationMessages, TranslationMessages } from '../../app/config/i18n';
import { store as fixturesStore } from '../../mocks/store';
import createReducer, { GlobalState } from '../../app/config/reducers';
import { ResponsiveThemeProvider } from '../../app/providers/responsiveThemeProvider';
import { CommonQuery } from '../../app/providers/commonQuery';
import { fillCommonQueryWithUser } from './commonQuery';

export const PLACEHOLDER_TEST_ID = 'content';
export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = `/${DEFAULT_LOCALE}`) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  history.push = jest.fn();
  return {
    history,
    pushSpy: history.push as jest.Mock,
  };
};

export interface ContextData {
  router?: {
    url?: string;
    routePath?: string;
    history?: MemoryHistory;
    children?: ReactNode;
  };
  store?: GlobalState | ((draft: GlobalState) => void);
  messages?: TranslationMessages;
  relayEnvironment?: RelayMockEnvironment | ((env: RelayMockEnvironment) => void);
}

export interface ProvidersWrapperProps {
  children: ReactNode;
  context?: ContextData;
}

export const ProvidersWrapper = ({ children, context = {} }: ProvidersWrapperProps) => {
  const { router = {}, messages } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory: MemoryHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  const store = (() => {
    if (!context?.store) return fixturesStore;
    if (typeof context.store === 'function') return produce(fixturesStore, context.store);
    return context.store;
  })();

  const relayEnvironment = (() => {
    if (context?.relayEnvironment === undefined) {
      const env = createMockEnvironment();
      fillCommonQueryWithUser(env);
      return env;
    }
    if (typeof context.relayEnvironment === 'function') {
      const mockEnvironment = createMockEnvironment();
      context.relayEnvironment(mockEnvironment);
      return mockEnvironment;
    }
    return context.relayEnvironment;
  })();

  return (
    <HistoryRouter history={routerHistory}>
      <HelmetProvider>
        <ResponsiveThemeProvider>
          <RelayEnvironmentProvider environment={relayEnvironment}>
            <CommonQuery>
              <IntlProvider {...intlProviderMockProps}>
                <Provider store={createStore(createReducer(), produce(store, identity))}>
                  <Routes>
                    <Route path={routePath} element={children}>
                      {router.children}
                    </Route>
                  </Routes>
                </Provider>
              </IntlProvider>
            </CommonQuery>
          </RelayEnvironmentProvider>
        </ResponsiveThemeProvider>
      </HelmetProvider>
    </HistoryRouter>
  );
};

export function makeContextRenderer<T>(
  component: (props: T | Record<string, never>) => ReactElement,
  baseContext: ContextData = {}
) {
  return (props?: T, context?: ContextData) => {
    return render(component(props ?? {}), {
      wrapper: ({ children }) => (
        <ProvidersWrapper context={{ ...baseContext, ...context }}>{children}</ProvidersWrapper>
      ),
    });
  };
}

export function makePropsRenderer<T>(component: (props: T | Record<string, never>) => ReactElement) {
  return (props?: T) => render(component(props ?? {}));
}

// using `screen.getByText(matchTextContent('hello world'))` will match <div>hello <span>world</span></div>
export const matchTextContent = (text: string | RegExp) => (_: unknown, node: Element | null | undefined) => {
  const hasText = (node: Element) => {
    const isPlainString = typeof text === 'string';
    const nodeContent = node.textContent ?? '';
    return isPlainString ? nodeContent === text : new RegExp(text).test(nodeContent);
  };
  const childrenDontHaveText = Array.from(node?.children ?? []).every((child) => !hasText(child));
  return Boolean(node && hasText(node) && childrenDontHaveText);
};

export const connectionFromArray = <T extends Record<string, any>>(arr: T[] = []) => {
  if (!arr || arr.length === 0) {
    return {
      edges: [],
      totalCount: 0,
      count: 0,
      endCursorOffset: 0,
      startCursorOffset: 0,
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
    };
  }

  return {
    edges: arr.map((node) => ({ node })),
    totalCount: arr.length,
    count: arr.length,
    endCursorOffset: arr.length,
    startCursorOffset: 0,
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
  };
};

export const packHistoryArgs = (pathname: string, other = {}) => [
  expect.objectContaining({ pathname, ...other }),
  undefined,
];
