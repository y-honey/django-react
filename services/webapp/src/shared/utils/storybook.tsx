import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { produce } from 'immer';
import { createReducer, Reducer } from '@reduxjs/toolkit';
import { Story } from '@storybook/react';
import { createMockEnvironment, RelayMockEnvironment } from 'relay-test-utils';
import { RelayEnvironmentProvider } from 'react-relay';
import { Suspense } from 'react';
import { createMemoryHistory } from 'history';
import { Route, Routes, unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
import { Elements } from '@stripe/react-stripe-js';
import { store } from '../../mocks/store';
import { GlobalState } from '../../app/config/reducers';
import { ActiveSubscriptionContext } from '../../routes/finances/activeSubscriptionContext/activeSubscriptionContext.component';
import { stripePromise } from '../services/stripe';
import { ContextData, ProvidersWrapper } from './testUtils';

export const withRedux =
  (initialState: GlobalState | ((draft: GlobalState) => void) = store, reducer: Reducer | null = null) =>
  (StoryComponent: Story) => {
    const state = typeof initialState === 'function' ? produce(store, initialState) : initialState;
    const createdStore = createStore(reducer ?? createReducer(state, {}));
    return (
      <Provider store={createdStore}>
        <StoryComponent />
      </Provider>
    );
  };

export const withRelay =
  (environmentArg?: RelayMockEnvironment | ((env: RelayMockEnvironment) => void)) => (StoryComponent: Story) => {
    const environment = (() => {
      if (typeof environmentArg === 'function') {
        const envToMutate = createMockEnvironment();
        environmentArg(envToMutate);
        return envToMutate;
      }
      if (environmentArg !== undefined) {
        return environmentArg;
      }
      return createMockEnvironment();
    })();

    return (
      <RelayEnvironmentProvider environment={environment}>
        <StoryComponent />
      </RelayEnvironmentProvider>
    );
  };

export const withProviders = (context?: ContextData) => (StoryComponent: Story) => {
  return (
    <ProvidersWrapper context={context}>
      <StoryComponent />
    </ProvidersWrapper>
  );
};

export const withSuspense = (StoryComponent: Story) => (
  <Suspense fallback={<span>Loading</span>}>
    <StoryComponent />
  </Suspense>
);

export const withActiveSubscriptionContext = (StoryComponent: Story) => {
  const routerHistory = createMemoryHistory({ initialEntries: ['/'] });
  return (
    <HistoryRouter history={routerHistory}>
      <Routes>
        <Route element={<ActiveSubscriptionContext />}>
          <Route
            index
            element={
              <Elements stripe={stripePromise}>
                <Suspense fallback={null}>
                  <StoryComponent />
                </Suspense>
              </Elements>
            }
          />
        </Route>
      </Routes>
    </HistoryRouter>
  );
};
