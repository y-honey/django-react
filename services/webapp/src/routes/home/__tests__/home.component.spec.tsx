import { screen } from '@testing-library/react';
import { Home } from '../home.component';
import { render } from '../../../tests/utils/rendering';

describe('Home: Component', () => {
  const Component = () => <Home />;

  it('should display welcome message', () => {
    render(<Component />);
    expect(screen.getByText('Welcome!')).toBeInTheDocument();
  });
});
