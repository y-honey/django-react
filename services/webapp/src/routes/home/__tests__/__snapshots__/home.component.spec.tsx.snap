// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home: Component should render correctly 1`] = `
.c1 {
  font-family: OpenSans;
  font-weight: bold;
  color: #000000;
}

.c3 {
  font-family: OpenSans;
  font-weight: bold;
  color: #000000;
}

.c4 {
  padding: 8px 16px;
  border: 1px solid #80809B;
  color: #42f272;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c2 {
  width: 100px;
  margin-bottom: 20px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    Hello world!
  </h1>
  <svg
    class="c2"
  >
    logo.svg
  </svg>
  <select
    class=""
  >
    <option
      value="en"
    >
      en
    </option>
    <option
      value="pl"
    >
      pl
    </option>
  </select>
  <h2
    class="c3"
  >
    Register
  </h2>
  <form
    class=""
  >
    <div
      class=""
    >
      <input
        class=""
        name="email"
        placeholder="Email"
        type="email"
      />
      <p
        class=""
      />
    </div>
    <div
      class=""
    >
      <input
        class=""
        name="password"
        placeholder="Password"
      />
      <p
        class=""
      />
    </div>
    <button
      class="c4"
      type="submit"
    >
      Signup
    </button>
  </form>
  <h2
    class="c3"
  >
    Login
  </h2>
  <form
    class=""
  >
    <div
      class=""
    >
      <input
        class=""
        name="email"
        placeholder="Email"
        type="email"
      />
      <p
        class=""
      />
    </div>
    <div
      class=""
    >
      <input
        class=""
        name="password"
        placeholder="Password"
      />
      <p
        class=""
      />
    </div>
    <button
      class="c4"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
